consensus/src/consensus_observer/observer/consensus_observer.rs

```diff
@@ -595,12 +635,48 @@ impl ConsensusObserver {
         let pending_block = self
             .ordered_block_store
             .lock()
-            .get_ordered_block(commit_decision.epoch(), commit_decision.round());
+            .get_observed_ordered_block(commit_decision.epoch(), commit_decision.round());
 
         // Process the pending block
         if let Some(pending_block) = pending_block {
-            // If all payloads exist, add the commit decision to the pending blocks
-            if self.all_payloads_exist(pending_block.blocks()) {
+            // Check if all block dependencies exist for the pending block
+            let block_dependencies_exist = match pending_block {
+                ObservedOrderedBlock::Ordered(ordered) => {
+                    if self.get_execution_pool_window_size().is_some() {
+                        // Log an error and return false (execution pool is enabled)
+                        error!(
+                            LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                                "Execution pool is enabled, but found a pending ordered block: {:?}",
+                                ordered.proof_block_info()
+                            ))
+                        );
+                        false
+                    } else {
+                        self.all_payloads_exist(&ordered)
+                    }
+                },
+                ObservedOrderedBlock::OrderedWithWindow(ordered_with_window) => {
```

Comment:
There's an edge case that needs to be addressed (not sure if here, but somewhere).

In consensus, when rebuilding the block store, blocks that are <= the last committed block are created without windows. It's guaranteed these blocks will not be executed. We need to be able to extend this guarantee to consensus observer. I wonder if the best way is to create a special "locally unavailable" window variant?

cc @hariria

Reply:
Good to know! I'll follow up on this once we come to actually execute the block in consensus observer.